// @grafana/toolkit/src/e2e/pages/dashboards/createDashboardPage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var pageInfo_1 = require("../../pageInfo");
var dashboardPage_1 = require("./dashboardPage");
var dashboardSettingsPage_1 = require("./dashboardSettingsPage");
var saveDashboardModal_1 = require("./saveDashboardModal");
var dashboardsPage_1 = require("./dashboardsPage");
var confirmModal_1 = require("../modals/confirmModal");
exports.createDashboardPage = new pageInfo_1.TestPage({
    url: '/dashboard/new',
    pageObjects: {
        addQuery: 'Add Query CTA button',
        saveDashboard: 'Save dashboard navbar button',
    },
});
exports.createEmptyDashboardPage = function (page, dashboardTitle) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
    return tslib_1.__generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, exports.createDashboardPage.init(page)];
            case 1:
                _a.sent();
                return [4 /*yield*/, exports.createDashboardPage.navigateTo()];
            case 2:
                _a.sent();
                return [4 /*yield*/, exports.createDashboardPage.pageObjects.saveDashboard.click()];
            case 3:
                _a.sent();
                return [4 /*yield*/, saveDashboardModal_1.saveDashboardModal.init(page)];
            case 4:
                _a.sent();
                return [4 /*yield*/, saveDashboardModal_1.saveDashboardModal.expectSelector({ selector: 'save-dashboard-as-modal' })];
            case 5:
                _a.sent();
                return [4 /*yield*/, saveDashboardModal_1.saveDashboardModal.pageObjects.name.enter(dashboardTitle)];
            case 6:
                _a.sent();
                return [4 /*yield*/, saveDashboardModal_1.saveDashboardModal.pageObjects.save.click()];
            case 7:
                _a.sent();
                return [4 /*yield*/, saveDashboardModal_1.saveDashboardModal.pageObjects.success.exists()];
            case 8:
                _a.sent();
                return [4 /*yield*/, dashboardPage_1.dashboardPage.init(page)];
            case 9:
                _a.sent();
                return [2 /*return*/, dashboardPage_1.dashboardPage];
        }
    });
}); };
exports.cleanDashboard = function (page, dashboardTitle) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
    var dashboardsPage;
    return tslib_1.__generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                dashboardsPage = dashboardsPage_1.dashboardsPageFactory(dashboardTitle);
                return [4 /*yield*/, dashboardsPage.init(page)];
            case 1:
                _a.sent();
                return [4 /*yield*/, dashboardsPage.navigateTo()];
            case 2:
                _a.sent();
                return [4 /*yield*/, dashboardsPage.pageObjects.dashboard.exists()];
            case 3:
                _a.sent();
                return [4 /*yield*/, dashboardsPage.pageObjects.dashboard.click()];
            case 4:
                _a.sent();
                return [4 /*yield*/, dashboardPage_1.dashboardPage.init(page)];
            case 5:
                _a.sent();
                return [4 /*yield*/, dashboardPage_1.dashboardPage.pageObjects.settings.click()];
            case 6:
                _a.sent();
                return [4 /*yield*/, dashboardSettingsPage_1.dashboardSettingsPage.init(page)];
            case 7:
                _a.sent();
                return [4 /*yield*/, dashboardSettingsPage_1.dashboardSettingsPage.pageObjects.deleteDashBoard.click()];
            case 8:
                _a.sent();
                return [4 /*yield*/, confirmModal_1.confirmModal.init(page)];
            case 9:
                _a.sent();
                return [4 /*yield*/, confirmModal_1.confirmModal.pageObjects.delete.click()];
            case 10:
                _a.sent();
                return [4 /*yield*/, confirmModal_1.confirmModal.pageObjects.success.exists()];
            case 11:
                _a.sent();
                return [2 /*return*/];
        }
    });
}); };
